db_stress_tool/no_batched_ops_stress.cc

```diff
@@ -571,7 +571,13 @@ class NonBatchedOpsStressTest : public StressTest {
     keys.reserve(num_keys);
     std::vector<PinnableSlice> values(num_keys);
     std::vector<Status> statuses(num_keys);
-    ColumnFamilyHandle* cfh = column_families_[rand_column_families[0]];
+    // When Flags_use_txn is enabled, we also do a read your write check.
```

Comment:
I found it not easy to grasp the structure of TestMultiGet which is not your fault by adding a new nested layer but the original complexity we have in TestMultiGet. With that being said, it would be nice if you can leave a comment at the top of the test, kind of describing the structure the test. By structure, I mean :

TestMultiGet: 
- use txn or not;
  - if use txn, we are testing (1) WriteBatchWithIndex::MultiGetFromBatchAndD and (2)  read your own write. Both without committing the txn; 
- do consistency check or not
   - if do consistency check, we set snapshot in read options and compare Get() VS MultiGet()

Reply:
I agree this verification part becomes very chunky after the read-your-write checks are also squeezed in. I added some comments to help list out its structure.

I made the added verification part of the consistency check, which previously refers to `Get` and `MultiGet` be consistent, and now also included checking `MultiGet` is consistent with what the transaction has written if `FLAGS_use_txn` is true. When do consistency check, the test has to take a snapshot to make sure reads from `MultiGet` and `Get` are repeatable, regardless of either `FLAGS_use_txn` is enabled because even when the flag is enabled, the transaction only attempts to write in a 1 in 10 chances, most of the time, it still reads from the db. This continues to be the case, the added verification is orthogonal to it.

These are slightly different error messages, I'm not sure how much variable naming/abstraction can help with deduplication,  vision decluttering.

Reply:
Sightly related to https://github.com/facebook/rocksdb/pull/11424#discussion_r1198166659 - is it correct that read-your-own-write should be expected to perform correctly even when ReadOptions' snapshot is not set? If it's true, should we also verify read-your-own-write when ReadOptions is not set with snapshot, where Get vs MultiGet consistency can't be verified (and not needed).

Reply:
That's a good point. Thank you for the suggestion, I added a read without snapshot consistency check for the keys with writes in the transaction. The consistency check is extracted out as a standalone function, although it's still a chunky function. But hopefully this helps better organize the test structure.

Reply:
Hey I am bit lost in the new artifacts (new variables and extracted function) for "a read without snapshot consistency check for the keys with writes in the transaction". I was merely suggesting something can be simple like below on top of you commit bb081d3bd840b823e65f9e081805cbb76523ac27. Let me know if I miss anything that have led you to use the new artifacts.

```
...
if (fault_fs_guard) {
  fault_fs_guard->DisableErrorInjection();
}

for (size_t i = 0; i < statuses.size(); ++i) {
  Status s = statuses[i];
  bool is_consistent = true;
  bool is_ryw_correct = true;
   ...
  // consistency check between Get and MultiGet
  if (do_consistency_check && !error_count && (s.ok() || s.IsNotFound())) {
    Status tmp_s;
    std::string value;
    if (use_txn) {
      tmp_s = txn->Get(readoptionscopy, cfh, keys[i], &value);
    } else {
      tmp_s = db_->Get(readoptionscopy, cfh, keys[i], &value);
    }
    if (!tmp_s.ok() && !tmp_s.IsNotFound()) {
      ...
    } else if (!s.ok() && tmp_s.ok()) {
      ...
    } else if (s.ok() && tmp_s.IsNotFound()) {
    ...
    } else if (s.ok() && value != values[i].ToString()) {
      ..
    }
  }
     
  // Check on read your own write in transaction
  if (is_consistent && use_txn) {
    assert(!expected_values.empty());
    if (!expected_values[i].has_value()) {
      continue;
    }
    ExpectedValue expected = expected_values[i].value();
    char expected_value[100];
    if (s.ok()) {
      Slice from_txn_slice(values[i]);
      if (ExpectedValueHelper::MustHaveNotExisted(expected, expected)) {
        fprintf(stderr,
                "MultiGet returned value different from what was written "
                "for key %s\n",
                keys[i].ToString(true).c_str());
        fprintf(stderr,
                "MultiGet returned ok, transaction has non-committed "
                "delete.\n");
        is_ryw_correct = false;
      } else {
        size_t sz = GenerateValue(expected.GetValueBase(), expected_value,
                                  sizeof(expected_value));
        Slice expected_value_slice(expected_value, sz);
        if (expected_value_slice.compare(from_txn_slice) == 0) {
          continue;
        }
        fprintf(stderr,
                "MultiGet returned value different from what was written "
                "for key %s\n",
                keys[i].ToString(true /* hex */).c_str());
        fprintf(stderr, "MultiGet returned value %s\n",
                from_txn_slice.ToString(true /* hex */).c_str());
        fprintf(stderr, "Transaction has non-committed value %s\n",
                expected_value_slice.ToString(true /* hex */).c_str());
        is_ryw_correct = false;
      }
    } else if (s.IsNotFound()) {
      if (ExpectedValueHelper::MustHaveExisted(expected, expected)) {
        size_t sz = GenerateValue(expected.GetValueBase(), expected_value,
                                  sizeof(expected_value));
        Slice expected_value_slice(expected_value, sz);
        fprintf(stderr,
                "MultiGet returned value different from what was written "
                "for key %s\n",
                keys[i].ToString(true).c_str());
        fprintf(stderr,
                "MultiGet returned not found, transaction has "
                "non-committed value.\n");
        is_ryw_correct = false;
      }
    }
  }
      

  if (!is_consistent) {
     fprintf(stderr, "TestMultiGet error: is_consistent is false\n");
     thread->stats.AddErrors(1);
     // Fail fast to preserve the DB state
     thread->shared->SetVerificationFailure();
     break;
   } else if (!is_ryw_correct) {
     // ...
     break;
   } else if (s.ok()) {
     // found case
     thread->stats.AddGets(1, 1);
   } else if (s.IsNotFound()) {
     // not found case
     thread->stats.AddGets(1, 0);
   } else if (s.IsMergeInProgress() && use_txn) {
     // With txn this is sometimes expected.
     thread->stats.AddGets(1, 1);
   } else {
     if (error_count == 0) {
      // errors case
      fprintf(stderr, "MultiGet error: %s\n", s.ToString().c_str());
      thread->stats.AddErrors(1);
    } else {
      thread->stats.AddVerifiedErrors(1);
    }
  }
}

if (readoptionscopy.snapshot) {
    db_->ReleaseSnapshot(readoptionscopy.snapshot);
}
...
```

My main suggestion is to extract the read-your-own-write check out of  `if (do_consistency_check..)` so the check happens even when `readoptionscopy.snapshot = db_->GetSnapshot();` is not called. `is_ryw_correct` might be needed if you want to print something similar to `  fprintf(stderr, "TestMultiGet error: is_consistent is false\n");` but for read your own write check

Reply:
I see what you mean. This was previously trying to do a `MultiGet` vs `Get` consistency check for keys in the transaction when no snapshot is set. I changed it to the structure you suggested. Thank you for the suggestion!

Reply:
> trying to do a MultiGet vs Get consistency check for keys in the transaction when no snapshot is set.

oh okay - then it seems like this can also be covered by the new structure cuz if the consistency check and read your write check pass, that means keys in the transactions are read correctly by MultiGet (by read your write check) and consistent with Get result (by consistency check). Did I miss anything here?

Reply:
`MultiGet` and `Get` consistency check is only done when `do_consistency_check` flag is true, currently done with a one in four chances, then a snapshot is taken. When `do_consistency_check` flag is set to false, no snapshot is taken, in that case, the consistency check will get bypassed, leaving the variable `is_consistent` true as its initialization value and goes directly to the read your write check that compares results from `MultiGet` to the locally remembered writes.